CustomSelectControl: Update to use a Popover component for rendering the internals

andrewserong opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Raised in https://github.com/WordPress/gutenberg/issues/36545#issuecomment-976447915, the `CustomSelectControl` component gets cut off when used in certain areas, such as in the global styles sidebar.

This PR explores using a Popover component for rendering the drop down items. The goal is to ensure that the dropdown items do not get cut off, and that we preserve accessibility and backwards compatibility of the component.

Kudos @aaronrobertshaw for the approach in this PR.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Open up the site editor and go to the global styles tab. Select block styles, and go to a block that opts in to Typography controls (e.g. the Heading block). Select from the font size dropdown and before this PR, you should see that the dropdown gets cut off visually. After this PR, the Popover component should try to ensure that the dropdown options are always visible, or at least can be scrolled to.

Check to ensure that keyboard navigation is unaffected.

You can also test in Storybook by running `npm run storybook:dev` and then navigate to: http://localhost:50240/?path=/story/components-customselectcontrol--long-labels

## To-do

* [x] Update styling to ensure the correct width is set for the dropdown
* [x] Update the position of where the Popover opens from, so that it's correctly aligned
* [x] Update the border of the Popover as it looks like it adds an extra grey border
* [x] ~Try using the `DropdownMenu` component instead of the `Popover` component~ (Reverted attempt to use `DropdownMenu` component due to accessibility issues)
* [x] Render popover position relative to the Container position and size 
* [ ] Place the Popover behaviour before an experimental prop (e.g. `__experimentalPopover`) so that we don't break backwards compatibility? Update: since we've managed to pretty well preserve the styling of the menu, I'm not sure this is worth it. Let me know if anyone thinks it's necessary, though!

## Screenshots <!-- if applicable -->

| Before | After |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/145519299-fc3dc08f-ae05-4e5b-8b4a-cbcc4d6e2f84.png) | ![Kapture 2021-12-10 at 15 27 28](https://user-images.githubusercontent.com/14988353/146133099-7e87e3dd-681f-4941-8db2-7c8baa958319.gif) |

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

andrewserong committed:
Try using Popover component in custom select control

andrewserong committed:
Remove anchorRect prop

aaronrobertshaw commented:
@youknowriad and @ciampo, if you have time, could you please give a general 👍 or 👎  on the approach taken here before we continue to refine the results as per the PR description todo list?

youknowriad commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -12,7 +12,15 @@ import { __, sprintf } from '@wordpress/i18n';
-import { Button, VisuallyHidden } from '../';
+import { Button, Popover, VisuallyHidden } from '../';
+
+const OptionList = ( { isOpen, children } ) => {
```
Instead of a custom implementation like that, maybe `DropdownMenu` is a component suited for this use-case (with all the focus management...)

andrewserong committed:
Switch to using the DropdownMenu component

andrewserong commented:
Thanks for the suggestion @youknowriad — I think `DropdownMenu` might be a good approach. It looks like it covers most of what we need, except for managing moving focus to the selected item when you open the dropdown menu. So, I think we might need to add some custom focus handling, there. However, aside from that, I think it seems to work pretty well, and would mean that we might be able to remove the `downshift` dependency, since it looks like this is the only component that uses it?

Here's a screengrab of where I got up to by the end of the day:

![Kapture 2021-12-13 at 17 31 42](https://user-images.githubusercontent.com/14988353/145763929-bd0cac55-98be-4cc1-93a2-04fc56467ed1.gif)

I'll keep chipping away at this, next up, I'll look into how we can set focus to the currently selected item when we open the dropdown menu.

aaronrobertshaw commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -112,71 +69,67 @@ export default function CustomSelectControl( {
+										style: item.style,
+									} }
+									onClick={ () => {
+										onSelectedItemChange( {
```
The readme for this component [states the `onChange` prop is optional](https://github.com/WordPress/gutenberg/blob/ae56763775c178aa4e020551c445fe607bd49931/packages/components/src/custom-select-control/README.md#onchange) so `onSelectedItemChange` here might not be a function. We hit this error when testing the recent changes in the Storybook as well.

aaronrobertshaw reviewed (commented):
Thanks for continuing to refine this @andrewserong 👌 

I took it for another premature test drive.  It's working well so far for me in the block editor.

In addition to the focus management items you've already mentioned, there were a few small issues that I believe will be ironed out as this evolves. In case it helps they were:

- The storybook example is currently broken - see [earlier comment](https://github.com/WordPress/gutenberg/pull/37272#discussion_r767468459)
- Styling, size, and position of the dropdown menu is quite different to before this update

<details>
    <summary> Before / After screenshots </summary>

| Before | After |
|---|---|
| <img width="280" alt="Screen Shot 2021-12-13 at 5 38 07 pm" src="https://user-images.githubusercontent.com/60436221/145772185-9caac560-0900-4c24-8204-82f8087e2180.png"> | <img width="437" alt="Screen Shot 2021-12-13 at 5 36 28 pm" src="https://user-images.githubusercontent.com/60436221/145772304-0e6ba01a-fb7f-4180-ba18-61ac0e280ff8.png"> |
| <img width="282" alt="Screen Shot 2021-12-13 at 5 37 59 pm" src="https://user-images.githubusercontent.com/60436221/145772274-07814027-ff90-4711-a782-a0585e39a12c.png"> | <img width="452" alt="Screen Shot 2021-12-13 at 5 37 38 pm" src="https://user-images.githubusercontent.com/60436221/145772230-bd4f93ed-cecd-4a11-bc8e-c88f7b95472c.png"> |
  
</details>

ciampo reviewed (commented):
Hey @andrewserong , thank you for exploring a solution to avoid the `CustomSelectControl` getting cut off. 

The main piece of feedback that I have is that, with the new approach in this PR, the semantics of the component are changing from being a `listbox` + `option` (like a native `select`) to being a `menu` + `group` + `menuitem`.

Given the fact that this component is a custom `select` component, I would argue that we should keep the same semantics — I'm not sure that `DropdownMenu` is the better choice in this regard, but I would like to hear also @diegohaz and @mirka 's opinion

andrewserong committed:
Explore overriding role attributes to semantically indicate that we are using approximating select/option

andrewserong committed:
Rename select role to listbox

andrewserong commented:
Thanks for the feedback @ciampo!

> The main piece of feedback that I have is that, with the new approach in this PR, the semantics of the component are changing from being a listbox + option (like a native select) to being a menu + group + menuitem.

That's a great point. I've had a go at updating this in commits (https://github.com/WordPress/gutenberg/pull/37272/commits/3e7af65f586a923cd210278a110ae91ce5a4b591 and https://github.com/WordPress/gutenberg/pull/37272/commits/48b92c188ca8499a01c8b22d1b02fed8c07a21e4) to see if we can get closer to the `listbox` and `option` semantics while using this component. I've done the following:

* Removed the `MenuGroup` nesting
* Moved where `role` is set within the `DropdownMenu` component so that a passed in `role` will be rendered, and then I've passed it a role of `listbox`
* Passed in a role of `option` to the `MenuItem` component

In terms of the rendered markup, it _looks_ like this gets us the `listbox` + `option` semantics (at least superficially):

#### Listbox

![image](https://user-images.githubusercontent.com/14988353/145937897-432d82cd-3e75-4da2-8263-aad86723dcf1.png)

#### Option

![image](https://user-images.githubusercontent.com/14988353/145937923-24e8cc32-d675-44ec-a850-754f45232dbb.png)

I haven't (yet) addressed @aaronrobertshaw's styling feedback or the other issues, as I wanted to confirm whether or not `DropdownMenu` is viable before polishing this PR too much. I _think_ in terms of behaviour the DropdownMenu component gets us most of what we need aside from setting focus on the currently selected item (while allowing us to potentially drop the `downshift` dependency). But, I'm curious to hear feedback on whether or not this usage of that component seems viable, or if it feels a little too hacky. In this case, I suppose ideally we could re-use behaviour of components and treat that behaviour as an implementation detail, while ensuring it has the appropriate semantics. But I'm very happy to defer to others, particularly given that I'm pretty fresh to looking at this component 😁

If `DropdownMenu` doesn't appear to be viable, do you think a custom `Popover` like [in my original commit](https://github.com/WordPress/gutenberg/pull/37272/commits/e6b9c88a8dea2f1835838a12d52c47217d19d65f) is a better option? Happy to explore any ideas 🙂

youknowriad commented:
Note that we also have the more generic `Dropdown` if we want to get rid of the menu semantics and apply our own.

ciampo commented:
@diegohaz any thoughts on this?

andrewserong committed:
Switch back to using Popover component, add useResizeObserver for setting correct width of dropdown

andrewserong committed:
Revert change to DropdownMenu component

andrewserong committed:
Remove unnecessary nesting

andrewserong commented:
Thanks for the feedback @youknowriad @ciampo — I've explored a couple of options today and ran into a couple of issues with both the `DropdownMenu` and `Dropdown` components, and I think our best bet is to revert back to using downshift + a custom `Popover`. This is because:

* Switching to `Dropdown` or `DropdownMenu` will involve a custom workaround for roles, which we could partially mitigate by passing a custom `role` attribute, but more importantly:
* It turns out `downshift` handles a lot of detailed keyboard interaction that wasn't immediately obvious to me, for example like a native `select` element, being able to type a letter and automatically move the selected option to the first option starting with that letter as described in the [listbox ARIA role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/listbox_role).
* Downshift also handles automatically focusing on the selected item within the popover, which we would have had to re-implement if using the `Dropdown` or `DropdownMenu` components.

The good news, is that it looks like we weren't too far off with the original addition of the `Popover` and downshift's keyboard handling appears to work nicely through that popover. To address the styling issues, I've added a `useResizeObserver` which grabs the width of the button and ensures that the options list matches the width of the button. This gets us much closer to preserving the original styling of the component. In the following screengrab, I'm using the keyboard to navigate through options, so it feels like things are working pretty well so far with the `Popover`:

![Kapture 2021-12-15 at 17 10 46](https://user-images.githubusercontent.com/14988353/146133099-7e87e3dd-681f-4941-8db2-7c8baa958319.gif)

Please let me know if there are any objections with this approach — the PR's still a work in progress (I only _just_ managed to get the styling looking pretty good before wrapping up for the day). For example, the Storybook demo styling is still pretty broken due to the centered position, but I'll fix up the positioning tomorrow:

![image](https://user-images.githubusercontent.com/14988353/146134022-24e40da3-628e-4558-9e98-dba1eca388b7.png)

I'm fairly sure that the addition of the Popover and preserving the rest of the existing component's behaviour is our best bet forward, but as always, very happy to hear other ideas 🙂

ciampo commented:
> I'm fairly sure that the addition of the Popover and preserving the rest of the existing component's behaviour is our best bet forward, but as always, very happy to hear other ideas 🙂

This also sounds like the most sensible (and least disruptive) approach. Feel free to ping again once you feel like the code is a good place for a review :)

youknowriad commented:
I'm fine with the plan but I do think it's a missed opportunity to consolidate our components and composability. The same behavior of CustomSelectControl might also be needed for ComboboxControl and potentially FormTokenField. why these components have different ways to handle popovers and keyboard navigation? They can all benefit from Dropdown and potentially hooks to handle the navigation.

diegohaz commented:
Semantically speaking, the select element is more similar to a combobox than to a dropdown menu. The dropdown menu trigger is a button. A button can only have a label. A select needs both a label and a value, and this is only possible with a few elements, like inputs, or elements with the `combobox` role. But, yeah, as @youknowriad mentioned, there's a lot of behavior that could be shared between these components. 

A custom select is also one of the most complicated widgets to implement (correctly). This is the best article I know about this topic: https://sarahmhigley.com/writing/select-your-poison/

For now, I agree with the plan. if we aren't going to use a library for that, like downshift or Ariakit (soon), I'd also consider using the native select whenever possible.

andrewserong commented:
Thanks for confirming the approach, folks!

> I'm fine with the plan but I do think it's a missed opportunity to consolidate our components and composability.

Agreed, initially I hoped we could consolidate at this stage. However, as Diego points out, there's a lot more complexity to implementing an accessible custom select component than I initially anticipated. After digging around in the keyboard navigation yesterday, I realised that `downshift` already handles all of the necessary `listbox` behaviour without us needing to do a larger refactor/rewrite, and it's good to contain scope creep on components when we can, it's very easy for the task to balloon!

I'm still keen for us to explore better composability / consolidation, but perhaps in the new year once the 5.9 work settles down (there's a good chunk of work for us to do on consolidating label spacing in components that we'd like to get to, too).

I'll ping again when this PR is ready for a review. Thanks everyone!

andrewserong committed:
Add anchorRef and merge refs with downshift's toggle button ref to allow calculating popover position based on the size of the button

andrewserong commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -81,6 +101,13 @@ export default function CustomSelectControl( {
+
+	const [
+		buttonResizeListener,
+		{ width: buttonWidth },
```
Note to self: this currently sets the width relative to the _button_ width, which is technically working nicely, but I _think_ on trunk, the width is set relative to the _parent_. This difference results in the following:

| Storybook on trunk | Storybook on this PR |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/146321921-be7958b4-7a11-4a9f-af62-07fded18c749.png) | ![image](https://user-images.githubusercontent.com/14988353/146321942-f0e9c0a5-2fd0-4ceb-8de0-4f6f32b269cd.png) |

Moving this up a level to get the width of the parent is probably the next step, we then might need to switch the position of the Popover to use an `anchorRect` or a `getAnchorRect` instead of `anchorRef` to set the right position relative to the width if the popover doesn't line up correctly. I'll have a play tomorrow.

andrewserong replied:
Resolved, I moved the `anchorRef` and width calculation up to the container in https://github.com/WordPress/gutenberg/pull/37272/commits/21c2f016349ad3c65284e15a1399782063f18e0f which solved most things.

andrewserong committed:
Move the anchorRef up to the container, calculate width factoring in the border width

andrewserong reviewed (commented):
Alrighty, I believe this is now ready for a proper review (CC: @ciampo). I've updated it to calculate the width for the menu via the width of the container, which I believe gets us pretty close to maintaining the same positioning / width as on `trunk`. If the animation of the popover feels like too much, we can switch it off with an `animate={ false }`. Here's a GIF of the current state within the editor (testing with the typography on the paragraph block is the most prominent place to test this):

![Kapture 2021-12-17 at 15 51 49](https://user-images.githubusercontent.com/14988353/146491007-b3b30a24-98a8-4e42-bbb5-8c7f54e4c710.gif)

A quick link to the Storybook demo after running `npm run storybook:dev` is: http://localhost:50240/?path=/story/components-customselectcontrol--long-labels

Let me know if you'd like to see any changes!

andrewserong committed:
Add changelog entry

andrewserong commented:
Rebased to fix conflicts with changelog.

ciampo commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -81,6 +101,18 @@ export default function CustomSelectControl( {
+		{ width: containerWidth },
+	] = useResizeObserver();
+	const widthMinusBorder =
+		containerWidth >= 2 ? containerWidth - 2 : containerWidth;
```
I wonder if there's a way to get the `border-width` of the `Popover` instead of hardcoding it. Maybe we could get a `ref` of the `Popover` and get the `computedStyle`s ?

ciampo replied:
Or maybe we could use a different value of `box-sizing` so that we don't need to take the border width into account at all?

andrewserong replied:
This one was slightly tricky because the border is set on `.components-popover__content` which is a child of the `Popover` container, so a `ref` doesn't quite get us there. Also, because `.components-popover__content` is a _parent_ of the element we're attaching `menuProps` to, `box-sizing` doesn't help us either — the border is set on the parent so the width of the child cannot take it into account. Also, I wanted to avoid overriding Popover classes directly where we can.

I _think_ the neatest thing to do here is to actually skip the JS calculation that added in `2px` and instead set a negative horizontal margin that references `$border-width` so we're still matching a variable instead of using a magic number (I've seen this in a couple of other places in the repo, too). I've added that in this commit: https://github.com/WordPress/gutenberg/pull/37272/commits/27c0cc2bb5b9cf844474453f6c4144c2a571c088

Happy to continue to iterate there if you can think of a better way to handle it, but 🤞 this makes things a bit neater.

ciampo commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -97,20 +129,36 @@ export default function CustomSelectControl( {
 		className: 'components-custom-select-control__menu',
 		'aria-hidden': ! isOpen,
+		style: {
+			minWidth: widthMinusBorder,
```
Why are we setting the min width, instead of the width itself?

andrewserong replied:
Similar to one of [the other comments](https://github.com/WordPress/gutenberg/pull/37272#discussion_r772072820), I thought that `minWidth` would make sense when I was calculating width from the Button / toggle width, so that the menu could expand further than the button. Now that we're calculating from the parent container, `width` makes more sense, so I've updated to `width` instead 👍

ciampo commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -125,58 +173,51 @@ export default function CustomSelectControl( {
-					) ) }
-			</ul>
+			<OptionList isOpen={ isOpen } anchorRef={ anchorRef }>
+				<ul { ...menuProps }>
+					{ isOpen &&
```
I wonder if the `ul` should instead be wrapped around `OptionList`. That would also avoid the `isOpen &&` check 

i.e

```jsx
<ul { ...menuProps }>
  <OptionList isOpen={ isOpen } anchorRef={ anchorRef }>
    { items.map( ... ) }
  </OptionList>
</ul>
```

andrewserong replied:
The `isOpen` check is slightly odd, but unfortunately I don't think we can move `OptionList` to be a child of the `ul`, as then the `Popover` divs would be direct descendants of the `ul` element (where only `li` / script tags [are allowed](https://html.spec.whatwg.org/multipage/grouping-content.html#the-ul-element)).

ciampo commented on packages/components/src/custom-select-control/style.scss:
```diff
@@ -41,18 +41,13 @@
-
 	max-height: 400px;
-	min-width: 100%;
+	min-width: 130px;
```
Why do we need a hardcoded px value here?

andrewserong replied:
I don't think we need it any more — I added it when I was calculating width based on the button instead of the container, where it seemed like a minimum width might be helpful. Now that we're calculating from the container, I think just going with the `width` value and not worrying about `min-width` makes more sense 👍

ciampo reviewed (commented):
Hey @andrewserong , thank you for iterating!

I think we're definitely moving in the right direction! Unfortunately, I won't be able to review again before the first week of January — hope that's ok!

andrewserong committed:
Move width calculation to negative margin, switch to width instead of minWidth

andrewserong commented:
Thanks for the review @ciampo!

> Unfortunately, I won't be able to review again before the first week of January — hope that's ok!

No worries, I'll be AFK for a couple of weeks from the end of this week, so if this PR hasn't been merged by then, feel free to merge it if it looks okay, otherwise I can continue on with it when I get back.

aaronrobertshaw commented on packages/components/src/custom-select-control/style.scss:
```diff
@@ -41,18 +41,12 @@
 	overflow: auto;
 	padding: 0;
-	position: absolute;
+	margin: 0 -$border-width; // Allow width value to include the Popover border width.
```
```suggestion
	margin: 0 (-$border-width); // Allow width value to include the Popover border width.
```

I believe the Sass compiler treats `-$border-width` like a subtraction operation. 

So in the original code `margin: 0 -$border-width;` results in `margin: -1px;` not `margin: 0 -1px`. Given the comment only mentions the width I assume this is unintended.

andrewserong replied:
Great catch, thanks Aaron! Yes, the negative margin is only intended for the horizontal margin. It's to reduce the width of the menu by 2px to account for the border of the parent, so that the popover lines up perfectly with the toggle button. I'll see if I can squeeze in a bit of time today to tweak it.

aaronrobertshaw reviewed (commented):
Thanks for iterating on this one @andrewserong 🙇 

I've tested again after the latest changes and I think it is getting close now.

✅ Storybook: Popover displays and functions as expected
✅ Site editor: Popover displays and is not cut-off in Global Styles panel
✅ Block editor: Popover renders
✅ Keyboard navigation: Works well in both Site and Block editors

❓ Styling: There are some issues with scrollbars showing when they shouldn't 

Everywhere I tested the updated `CustomSelectControl` displayed extraneous scrollbars. I believe this is in part due to a small issue with the negation of the `$border-width` Sass variable. Quickly hacking around in dev tools, setting the popover menu's margin to `0` sorted out the scrollbars. I'm lacking some of the history behind the attempt to account for border-width so I'm probably missing something.

I've left an inline comment and suggestion regarding the Sass variable negation.

<details>
<summary>Screenshots</summary>
<img width="250" alt="Screen Shot 2021-12-22 at 3 17 10 pm" src="https://user-images.githubusercontent.com/60436221/147039583-9e0b9876-d767-4afb-828e-1ada5466decd.png">
<img width="250" alt="Screen Shot 2021-12-22 at 3 06 42 pm" src="https://user-images.githubusercontent.com/60436221/147039594-ba84e870-96bf-47e1-89bc-65bf36c6e829.png">
<img width="400" alt="Screen Shot 2021-12-22 at 3 08 53 pm" src="https://user-images.githubusercontent.com/60436221/147039588-fb637a01-0d3d-47a3-b5e3-146f8c2f02c3.png">
<img width="400" alt="Screen Shot 2021-12-22 at 3 08 33 pm" src="https://user-images.githubusercontent.com/60436221/147039590-acf09470-8f98-4156-96c7-1f67487250dd.png">
<img width="400" alt="Screen Shot 2021-12-22 at 3 08 27 pm" src="https://user-images.githubusercontent.com/60436221/147039592-e0ca0d4a-dd33-40a9-b3fc-339ec786402f.png">

</details>

andrewserong committed:
Try to fix scrollbar issues

andrewserong commented:
Thanks again for reviewing @aaronrobertshaw! I _thought_ I'd come up with a simple fix for the scrollbar issue by switching off overflow for the Popover (pushed in https://github.com/WordPress/gutenberg/pull/37272/commits/f261d3b7b6d6e572e86dddd1f2d779da5bfa50ae) that _kind of_ works. An additional problem for us with the scrollbars is with how the Menu area behaves when the vertical room is limited and how the Popover behaves. At the moment, _both_ expect to scroll the content, and I ran out of time to tease apart which should "really" scroll the content, as choosing either one introduced subtle bugs. The way I've left it at the moment, is that they're switched off on the Popover, which superficially looks better:

<details>
<summary>Screenshots</summary>

| Before | After |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/147199697-481d8307-c525-46b9-853d-27bff3fba0a8.png) | ![image](https://user-images.githubusercontent.com/14988353/147199712-023ef4b7-4974-4d18-b386-924b51e62b52.png) |

</details>

However, if you shorten your browser window, and then use the keyboard to scroll up and down between items, the selected item will scroll off the edge of the screen. Not quite what we want!

I _think_ what we want is for the actual height of the popover to be limited to the vertical edge of the screen, without scrolling at all, and for the Menu to be the true source of the scrollbar, but I didn't quite figure it out before wrapping up for the year.

If anyone would like to have a go at fixing it, feel free to! Otherwise, I'll take another look at it with fresh 👀 when I'm back in a couple of weeks 🙂

ciampo reviewed (commented):
Not sure if this solves the problem at the root, but while playing around with Storybook, I noticed that the scrollbar (at least in the Storybook example) is caused by the "huge" item — in particular because its font size is quite large (`200%`), while the `line-height` is fixed (`28px`). This causes the item to text to overflow its item's height, and the scrollbar to appear:

![Screenshot 2022-01-08 at 11 34 26](https://user-images.githubusercontent.com/1083581/148641024-b296c141-722d-4519-84fc-25adfd88a4f5.png)

Disabling the `line-height` rule allows the `li` item (and therefore the `ul` parent) to resize to the correct height, and thus avoids the scrollbar

![Screenshot 2022-01-08 at 11 36 49](https://user-images.githubusercontent.com/1083581/148641077-56f556d1-d5c1-4804-841c-0f67b732c934.png)

I wonder if we should rewrite some of the CSS so that:
- the `line-height` is expressed in unitless values, becoming proportional to the font size
- the `li` item has a min-height, and it otherwise grows as needed by its contents

andrewserong commented:
Thanks for all the earlier feedback on this now very old PR. Since this is stale, and it looks like it'll be superseded by #41726 which refactors the component, I'll close this one out.
